### DISCUSSION TITLE
Add request Header in middleware

### DISCUSSION THREAD

**Discussion by kienerj** [fastapi/fastapi#2921] [Questions]
How can I add a header to all request in a middleware? It seems that request.headers is actually immutable is there trick to to it anyway?

If not how else can I pass a piece of information to all requests? (getting that piece of information might require a redirect back and forth)

**Comment by ec2ainun** [comment]
@kienerj 
i am not quite understand what's your intent is,

if you want to set Header on each respond, take a look at [this][3]
you can append it on response.Header

if you want to get header value in request, you can use [header param][1]
if you want to set header value in respond, you can use [respond header][2]

```py
         yourHeaderObject = {}
         yourReturnData = {}
         yourMimetype = "application/json"
         return Response(content=yourReturnData, status_code=200, headers=yourHeaderObject, media_type=yourMimetype)
```



[1]: https://fastapi.tiangolo.com/tutorial/header-params
[2]: https://fastapi.tiangolo.com/advanced/response-headers
[3]: https://fastapi.tiangolo.com/tutorial/middleware

**Reply by kienerj** [reply]
I want to add a header to the incoming request before passing it on the target API method, On the request, not on the response.

Doesn't necessarily need to be a header, could also be via other means of passing the information to the API method.

**Reply by Mause** [reply]
You can create a new request with the new header?

**Reply by kienerj** [reply]
How do I copy everything relevant from the original request to this new request? (I'm new to fastapi /starlette etc)

**Reply by kienerj** [reply]
I now found in starlette docs that one can add anything to request.state.  So I can add my information there and then retrieve it again inside API methods (with the downside of needing request as a method param on each api method)
